### PR TITLE
MNT: make Axes.cla an alias for Axes.clear in all cases

### PR DESCRIPTION
**Pull request by lijake8**
## PR Summary
We (myself and @neil-gurnani) consolidated all implementations of cla/clear under a clear method (the clear functionality is implemented in clear, and cla calls clear for the sake of minimizing code duplication and improving maintainability). We added 2 tests ensuring subclasses do not override cla. Existing tests for cla were left as-is; since cla calls clear, coverage is not compromised.

Every subclass has both cla and clear, since it was mentioned in #22738 that cla is still too widely used to be fully deprecated as of now. We added appropriate docstrings/comments for all cla/clear methods.

We left all original '_api.deprecated' decorators as-is but did not add any new decorators (please let us know if this is needed).

Finally, please let us know if there's any issue with this PR we should address. This was our first PR and we really enjoyed working on it. 

Thanks

## PR Checklist

**Tests and Styling**
- [x] Has pytest style unit tests (and `pytest` passes).
- [x] Is [Flake 8](https://flake8.pycqa.org/en/latest/) compliant (install `flake8-docstrings` and run `flake8 --docstring-convention=all`).

**Documentation**
- [N/A] New features are documented, with examples if plot related.
- [N/A] New features have an entry in `doc/users/next_whats_new/` (follow instructions in README.rst there).
- [N/A] API changes documented in `doc/api/next_api_changes/` (follow instructions in README.rst there).
- [N/A] Documentation is sphinx and numpydoc compliant (the docs should [build](https://matplotlib.org/devel/documenting_mpl.html#building-the-docs) without error).

### CODE REVIEW

File: examples/misc/custom_projection.py
Comment (greglucas) [inline examples/misc/custom_projection.py:78]: Examples probably don't need to define cla() at all.

File: lib/matplotlib/projections/geo.py
Comment (greglucas) [inline lib/matplotlib/projections/geo.py:61]: I think the `cla()` here can be inherited from the baseclass? So make sure that `Axes` has defined `cla()` to call `self.clear()`, which will then call your redefined `self.clear()` above. I'm pretty sure you already have this, so removing this method should be fine.

File: lib/matplotlib/projections/geo.py
Comment (greglucas) [inline lib/matplotlib/projections/geo.py:501]: This can be removed too.

File: lib/matplotlib/projections/polar.py
Comment (greglucas) [inline lib/matplotlib/projections/polar.py:384]: Since this was deprecated in 3.4, it can be removed in 3.6 which is our next release. So, you can remove this here if you want. You'd need to add a doc api_change_note for removals indicating that it has been removed too.

File: lib/matplotlib/projections/polar.py
Comment (greglucas) [inline lib/matplotlib/projections/polar.py:684]: Remove this as well if you want along with an API-change note.

File: lib/matplotlib/projections/polar.py
Comment (greglucas) [inline lib/matplotlib/projections/polar.py:810]: Can remove this definition

File: lib/matplotlib/spines.py
Comment (greglucas) [inline lib/matplotlib/spines.py:228]: Remove with an API change removal note if you want as well.

File: lib/mpl_toolkits/axes_grid1/axes_grid.py
Comment (greglucas) [inline lib/mpl_toolkits/axes_grid1/axes_grid.py:49]: I think this instance of `cla()` does need to be kept around because this class isn't inheriting from Axes.

File: lib/matplotlib/tests/test_axes.py
Comment (greglucas) [inline lib/matplotlib/tests/test_axes.py:474]:
```suggestion
    for klass in Axes.__subclasses__():
        # check that subclasses do not get redefined in our Axes subclasses
```
I think you want the "e" Axes instead of Axis?

**Review by greglucas — COMMENTED** [review]
Looks like a great start! I think there were a few places that you added in a definition of `cla()` into a subclass when it could be inherited from the superclass, so I'd suggest looking at what each of the classes is inheriting from and making sure that there is a cla() in the parent that can get used. Let me know if that doesn't make sense or if I'm mistaken too!

Also, if you want to add your friend as a contributor on the commit, you can use the "co-authored-by" feature in the commit message: https://docs.github.com/en/github-ae@latest/pull-requests/committing-changes-to-your-project/creating-and-editing-commits/creating-a-commit-with-multiple-authors

**Comment by lijake8** [comment]
> Looks like a great start! I think there were a few places that you added in a definition of `cla()` into a subclass when it could be inherited from the superclass, so I'd suggest looking at what each of the classes is inheriting from and making sure that there is a cla() in the parent that can get used. Let me know if that doesn't make sense or if I'm mistaken too!

@greglucas Thanks for the feedback and suggestions! I just amended the PR. Summary:
- The cla() methods are now removed from all axes subclasses. 
- The test case is also updated and (locally) is verified to be passing. 
- Relevant deprecations have been removed and logged in the doc/api/next_api_changes/removals folder. 

Please let me know if there's anything else I should do!

File: lib/mpl_toolkits/axes_grid1/parasite_axes.py
Comment (greglucas) [inline lib/mpl_toolkits/axes_grid1/parasite_axes.py:28]:
```suggestion
```
This is an interesting one because it looks like this isn't inheriting from Axes, but down below there is a class factory to create the ParasiteAxes from Axes, which I think means this does actually get a `cla()` from Axes when used in practice.

File: lib/mpl_toolkits/axes_grid1/parasite_axes.py
Comment (greglucas) [inline lib/mpl_toolkits/axes_grid1/parasite_axes.py:154]:
```suggestion
```

File: lib/mpl_toolkits/axisartist/floating_axes.py
Comment (greglucas) [inline lib/mpl_toolkits/axisartist/floating_axes.py:348]:
```suggestion
```

File: lib/matplotlib/pyplot.py
Comment (neil-gurnani) [reply lib/matplotlib/pyplot.py:1116]: Resolved, updated the `clear()` function in `pyplot.py` to use gca().cla().

File: lib/matplotlib/tests/test_figure.py
Comment (neil-gurnani) [reply lib/matplotlib/tests/test_figure.py:803]: Updated `lib/matplotlib/tests/test_figure.py` to remove the `test_cla_not_redefined()` for the reason specified.

File: lib/mpl_toolkits/axes_grid1/parasite_axes.py
Comment (neil-gurnani) [reply lib/mpl_toolkits/axes_grid1/parasite_axes.py:28]: Removing the `def cla(self)` because `ParasiteAxes` inherits from Axes with the class factory, as seen on lines 97-99 of `lib/mpl_toolkits/axes_grid1/parasite_axes.py`.

File: lib/mpl_toolkits/axes_grid1/parasite_axes.py
Comment (neil-gurnani) [reply lib/mpl_toolkits/axes_grid1/parasite_axes.py:154]: Removing the `def cla(self)` because `HostAxes` also inherits from Axes with the class factory, as seen on lines 239-242 of `lib/mpl_toolkits/axes_grid1/parasite_axes.py`. 
Note: line numbers specified are before removal of the function.

File: lib/mpl_toolkits/axisartist/floating_axes.py
Comment (neil-gurnani) [reply lib/mpl_toolkits/axisartist/floating_axes.py:348]: Removing the `def cla(self)` because `FloatingAxes` also inherits from Axes with the class factory, as seen on line 359-363 of `lib/mpl_toolkits/axisartist/floating_axes.py`.

File: lib/matplotlib/pyplot.py
Comment (greglucas) [reply lib/matplotlib/pyplot.py:1120]: I would just not add the `clear()` function to pyplot. So, don't touch this file at all in your changes in the PR.

It looks like your updates didn't get pushed up to the remote, I only see a new merge commit after your comments.

**Comment by lijake8** [comment]
Just pushed our updates. Unfortunately, we accidentally fetched from upstream before pushing and as a result there are some changes (particularly in pyplot) that are from upstream (the root matplotlib repo). Hopefully this is ok - if not, is there any way around this?

**Comment by neil-gurnani** [comment]
@greglucas Thanks for the feedback and suggestions! When trying to merge my changes to the current commit, I kept running into Failed to Push Refs due to stale info errors as shown [here](https://stackoverflow.com/questions/56191415/why-is-git-push-force-with-lease-failing-with-rejected-stale-info-even). Then I tried to "fetch the updates from upstream and rebase first" as outlined in [this source](fetch the updates from upstream and rebase first), which caused the merge from main commit you saw. 

Anyways, my partner (@lijake8) was able to amend our changes for this PR. 

Summary:

- Reverted `pyplot.py` file to the pre-PR version with no addition of a `clear()` function

- Removed the `test_cla_not_redefined()` function from the `test_figure.py` file.

- Removed the def cla(self) for `ParasiteAxes` , `HostAxes`, and `FloatingAxes` as these three inherit from the `Axes` parent class using the`ClassFactory` method. 

Please let me know if there's anything else I should do!

**Comment by lijake8** [comment]
FInally fixed the commits issue, and the relevant tests to the changes pass, but now the automated main tests aren't passing because
`lib/matplotlib/tests/test_pyplot.py::test_pyplot_up_to_date`
fails. What should we do? This will resolve after merging with the main branch, right?

File: lib/matplotlib/tests/test_figure.py
Comment (greglucas) [inline lib/matplotlib/tests/test_figure.py:799]: Hm... Is flake8 not getting run on tests? I'm pretty sure this line should be there. This file should also be left alone if you aren't touching it anymore.

File: lib/mpl_toolkits/axisartist/axislines.py
Comment (greglucas) [inline lib/mpl_toolkits/axisartist/axislines.py:504]:
```suggestion
        # gridlines need to be created before clear() since clear calls grid()
```
Might as well fix the typo while we're at it.

**Review by greglucas — APPROVED** [review]
Looks good to me! Just two very minor comments here. Then we just need to wait for a second reviewer before merging.

File: lib/matplotlib/tests/test_figure.py
Comment (lijake8) [reply lib/matplotlib/tests/test_figure.py]: Fixed

File: lib/mpl_toolkits/axisartist/axislines.py
Comment (lijake8) [reply lib/mpl_toolkits/axisartist/axislines.py]: Fixed

**Comment by lijake8** [comment]
@timhoffm @oscargus Do either of you mind reviewing this PR? Thanks!

**Review by timhoffm — APPROVED** [review]

**Comment by timhoffm** [comment]
Thanks @lijake8 and congratulations on your first contribution to Matplotlib! We hope to see you again.

**Comment by neil-gurnani** [comment]
@timhoffm My partner @lijake8 and I both worked on contributing to this issue together. 

He also added my name as a co-author as instructed by @greglucas above: 
"Also, if you want to add your friend as a contributor on the commit, you can use the "co-authored-by" feature in the commit message: https://docs.github.com/en/github-ae@latest/pull-requests/committing-changes-to-your-project/creating-and-editing-commits/creating-a-commit-with-multiple-authors"

Can you confirm that both myself @neil-gurnani (Neil Gurnani) and my partner (@lijake8) will be credited for this contribution?

Thank you for your guidance with this and your approval of our Pull Request!

**Comment by timhoffm** [comment]
> Can you confirm that both myself @neil-gurnani (Neil Gurnani) and my partner (@lijake8) will be credited for this contribution?

You’ve not followed the format that GitHub defines for co-author attribution. So on the technical side your account is not linked to the PR. OTOH your name is in the commit message, which states that you have contributed.

**Comment by tacaswell** [comment]
As Tim said, it looks like the commit message did not trigger the co-author logic in GH.  I suspect the issue is that GH is looking for the tags at the end, but you put it in the first line of the commit message (that normally says what it is you did).

On one hand I am sympathetic to wanting to get this fixed and get your name listed on GH as a contributor.  On the other hand I do not think we want to do a force-push to main to re-write history to fix the commit message (we _have_ force pushed to main in the past but it is my judgement that this does not cross the threshold to do that) and I do not think that adding an empty commit with your name on it is a good option either (it makes it really odd if anyone looks it up later and I think is a bad precedent to set). 

Given that the concept of co-authoring is a GH concept (rather than a pure git concept), it may be worth reaching out to github support about this.   At a minimum they can add a warning to their docs about where the line is searched for.

**Pull request merged**
